src/main/java/frc/robot/swerve/ModuleIOReal.java

```diff
@@ -0,0 +1,165 @@
+        inputs.driveMotorSupplyCurrentOverTime = driveSupplyChargeUsedCoulomb.get();
+        driveStatorChargeUsedCoulomb.update(inputs.driveMotorStatorCurrent);
+        inputs.driveMotorStatorCurrentOverTime = driveStatorChargeUsedCoulomb.get();
+        inputs.driveMotorPosition = driveMotor.getRotorPosition().getValue();
+        inputs.driveMotorVelocity = getVelocity();
+        inputs.driveMotorVelocitySetpoint = driveMotorVelocitySetpoint;
+
+        inputs.angleMotorSupplyCurrent = angleMotor.getSupplyCurrent().getValue();
+        inputs.angleMotorStatorCurrent = angleMotor.getStatorCurrent().getValue();
+        angleSupplyChargeUsedCoulomb.update(inputs.angleMotorSupplyCurrent);
+        inputs.angleMotorSupplyCurrentOverTime = angleSupplyChargeUsedCoulomb.get();
+        angleStatorChargeUsedCoulomb.update(inputs.angleMotorStatorCurrent);
+        inputs.angleMotorStatorCurrentOverTime = angleStatorChargeUsedCoulomb.get();
+        inputs.angleMotorPosition = angleMotor.getRotorPosition().getValue();
+        inputs.angleMotorVelocity = angleMotor.getVelocity().getValue();
+
+        inputs.angle = getAngle();
+        currentAngle = inputs.angle;
+
+        inputs.angleSetpoint = angleSetpoint;
+
+        inputs.moduleDistance = getModulePosition().distanceMeters;
+        inputs.moduleState = getModuleState();
+    }
+
+    @Override
+    public Rotation2d getAngle() {
+        return new Rotation2d(AngleUtil.normalize(Units.rotationsToRadians(angleMotor.getPosition().getValue())));
+    }
+
+    @Override
+    public void setAngle(Rotation2d angle) {
+        angleSetpoint = angle;
+        Rotation2d error = angle.minus(currentAngle);
+        angleControlRequest.withPosition(angleMotor.getPosition().getValue() + error.getRotations());
+        angleMotor.setControl(angleControlRequest);
+    }
+
+    @Override
+    public double getVelocity() {
+        return driveMotor.getVelocity().getValue();
+    }
+
+    @Override
+    public void setVelocity(double velocity) {
+        var angleError = angleSetpoint.minus(currentAngle);
+        velocity *= angleError.getCos();
+
+        driveMotorVelocitySetpoint = velocity;
+
+        velocityControlRequest.withVelocity(utils.units.Units.metersToRotations(
+                        velocity,
+                        SwerveConstants.WHEEL_DIAMETER / 2));
+        driveMotor.setControl(velocityControlRequest);
+    }
+
+    @Override
+    public SwerveModuleState getModuleState() {
+        return new SwerveModuleState(
+                utils.units.Units.rpsToMetersPerSecond(
+                        getVelocity(), SwerveConstants.WHEEL_DIAMETER/2)
+                , getAngle()
+        );
+    }
+
+    @Override
+    public SwerveModulePosition getModulePosition() {
+        return new SwerveModulePosition(
+                utils.units.Units.rpsToMetersPerSecond(
+                        driveMotor.getPosition().getValue(),
```

Comment:
Looks mixed again -- using distance function with a velocity argument.

Reply:
The calculation from rps to velocity is the same as rotations to meters.